serde-generate/src/ocaml.rs

```diff
@@ -0,0 +1,280 @@
+use crate::{
+    indent::{IndentConfig, IndentedWriter},
+    CodeGeneratorConfig,
+};
+use heck::SnakeCase;
+use include_dir::include_dir as include_directory;
+use serde_reflection::{ContainerFormat, Format, Named, Registry, VariantFormat};
+use std::{
+    collections::{BTreeMap},
+    io::{Result, Write},
+    path::PathBuf,
+};
+
+fn capitalize(s: &str) -> String {
+    let mut c = s.chars();
+    match c.next() {
+        None => String::new(),
+        Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
+    }
+}
+
+pub struct CodeGenerator<'a> {
+    config: &'a CodeGeneratorConfig,
+    libraries: Vec<String>,
+}
+
+struct OCamlEmitter<'a, T> {
+    out: IndentedWriter<T>,
+    generator: &'a CodeGenerator<'a>,
+}
+
+impl<'a> CodeGenerator<'a> {
+
+    pub fn new(config: &'a CodeGeneratorConfig) -> Self {
+        if config.c_style_enums {
+            panic!("OCaml does not support generating c-style enums");
+        }
+        Self {
+            config,
+            libraries: config.external_definitions.keys().map(|k| k.to_string()).collect::<Vec<_>>(),
+        }
+    }
+
+    pub fn output(&self, out: &mut dyn Write, registry: &Registry) -> Result<()> {
+        let mut emitter = OCamlEmitter {
+            out: IndentedWriter::new(out, IndentConfig::Space(2)),
+            generator: self,
+        };
+        emitter.output_preamble()?;
+        // todo: do a strongly connected components sort to regroup types that are cyclic
+        let mut i = 0;
+        let n = registry.len();
+        for (name, format) in registry {
+            let first = i == 0;
+            let last = i == n-1;
+            emitter.output_container(name, format, first, last)?;
+            i = i + 1;
+        }
+        Ok(())
+    }
+}
+
+impl<'a, T> OCamlEmitter<'a, T>
+where T: Write, {
+
+    fn output_preamble(&mut self) -> Result<()> {
+        for namespace in self.generator.libraries.iter() {
```

Comment:
To run a more complex example, I personally had to add `[@@@warning "-30"]` at the beginning of the file.

Reply:
Removing warning 30 and 42 are indeed the way to go, but I think they are already discarded in the last version of dune (at least they are discarded in ocamlc).
If it is not the case, it will be simpler to add it directly in the dune file.

Reply:
Mmm, at least that was the case for me and just installed dune (v3.0.3).